libraries/LittleFS/src/LittleFS.cpp

```diff
@@ -0,0 +1,133 @@
+    if (!_mounted) {
+        DEBUGV("LittleFSImpl::open() called on unmounted FS\n");
+        return FileImplPtr();
+    }
+    if (!path || !path[0] || (strlen(path) >= LFS_NAME_MAX)) {
+        DEBUGV("LittleFSImpl::open() called with invalid filename\n");
+        return FileImplPtr();
+    }
+    int flags = _getFlags(openMode, accessMode);
+    auto fd = std::make_shared<lfs_file_t>();
+    char lfsPath[LFS_NAME_MAX+1];
+    int rc = lfs_file_open(&_lfs, fd.get(), spiffsNameToLFS(path, lfsPath), flags);
+    if (rc < 0) {
+        DEBUGV("LittleFSDirImpl::openFile: rc=%d fd=%p path=`%s` openMode=%d accessMode=%d err=%d\n",
+               rc, fd.get(), path, openMode, accessMode, rc);
+        return FileImplPtr();
+    }
+    return std::make_shared<LittleFSFileImpl>(this, path, fd);
+}
+
+DirImplPtr LittleFSImpl::openDir(const char *path) {
+    if (!_mounted) {
+        return DirImplPtr();
+    }
+    auto dir = std::make_shared<lfs_dir_t>();
+    int rc = lfs_dir_open(&_lfs, dir.get(), "/");
+    if (rc < 0) {
+        DEBUGV("LittleFSImpl::openDir: path=`%s` err=%d\n", path, rc);
+        return DirImplPtr();
+    }
+    // Skip the . and .. entries
+    lfs_info dirent;
+    lfs_dir_read(&_lfs, dir.get(), &dirent);
+    lfs_dir_read(&_lfs, dir.get(), &dirent);
+
+    return std::make_shared<LittleFSDirImpl>(path[0]=='/'?(path+1):path, this, dir);
+}
+
+
+// Borrow the existing SPIFFS flash HAL
+extern int32_t spiffs_hal_write(uint32_t addr, uint32_t size, const uint8_t *src);
+extern int32_t spiffs_hal_erase(uint32_t addr, uint32_t size);
+extern int32_t spiffs_hal_read(uint32_t addr, uint32_t size, uint8_t *dst);
+
+int LittleFSImpl::lfs_flash_read(const struct lfs_config *c,
+    lfs_block_t block, lfs_off_t off, void *dst, lfs_size_t size) {
+    LittleFSImpl *me = reinterpret_cast<LittleFSImpl*>(c->context);
+    uint32_t addr = me->_start + (block * me->_blockSize) + off;
+    return spiffs_hal_read(addr, size, static_cast<uint8_t*>(dst));
+}
+
+int LittleFSImpl::lfs_flash_prog(const struct lfs_config *c,
+    lfs_block_t block, lfs_off_t off, const void *buffer, lfs_size_t size) {
+    LittleFSImpl *me = reinterpret_cast<LittleFSImpl*>(c->context);
+    uint32_t addr = me->_start + (block * me->_blockSize) + off;
+    const uint8_t *src = reinterpret_cast<const uint8_t *>(buffer);
+    return spiffs_hal_write(addr, size, static_cast<const uint8_t*>(src));
+}
+
+int LittleFSImpl::lfs_flash_erase(const struct lfs_config *c, lfs_block_t block) {
+    LittleFSImpl *me = reinterpret_cast<LittleFSImpl*>(c->context);
+    uint32_t addr = me->_start + (block * me->_blockSize);
+    uint32_t size = me->_blockSize;
+    return spiffs_hal_erase(addr, size);
+}
+
+int LittleFSImpl::lfs_flash_sync(const struct lfs_config *c) {
+    /* NOOP */
+    (void) c;
+    return 0;
+}
+
+
+// these symbols should be defined in the linker script for each flash layout
+#ifndef CORE_MOCK
+#ifdef ARDUINO
+#ifndef SPIFFS_MAX_OPEN_FILES
```

Comment:
Rather than use all of these SPIFFS symbols directly, I suggest implementing LITTLEFS versions of them, and have those implementations in terms of the SPIFFS ones. Example:
```
#define LITTLEFS_MAX_OPEN_FILES SPIFFS_MAX_OPEN_FILES

uint32_t _LITTLEFS_start = _SPIFFS_start;

inline int littlefs_hal_read(blah)
{
  return spiffs_hal_read(blah);
}

int LittleFSImpl::lfs_flash_read(const struct lfs_config *c,
    lfs_block_t block, lfs_off_t off, void *dst, lfs_size_t size) {
    LittleFSImpl *me = reinterpret_cast<LittleFSImpl*>(c->context);
    uint32_t addr = me->_start + (block * me->_blockSize) + off;
    return littlefs_hal_read(addr, size, static_cast<uint8_t*>(dst));
}

```

Put all of the definitions elsewhere in an isolated file, so that spiffs symbols are ever only referenced in that one place.
Why do this? The implementation is then better self-contained, which would help if we decide to drop spiffs in favor of littlefs.

Reply:
Refactored to this kind of setup in the latest push.